assets/js/upload.js

```diff
@@ -171,9 +171,121 @@ function setupImageUpload() {
     window.removeEventListener('beforeunload', beforeUnload);
   }
 
+  function createTagError(message) {
+    const buttonAfter = $('#tagsinput-save');
+    const errorElement = makeEl('span', { className: 'help-block tag-error' });
+
+    errorElement.innerText = message;
+    buttonAfter.insertAdjacentElement('beforebegin', errorElement);
+  }
+
+  function clearTagErrors() {
+    $$('.tag-error').forEach(el => el.remove());
+  }
+
+  const ratingsTags = ['safe', 'suggestive', 'questionable', 'explicit', 'semi-grimdark', 'grimdark', 'grotesque'];
+
+  // populate tag error helper bars as necessary
+  // return true if all checks pass
+  // return false if any check fails
+  function validateTags() {
+    const tagInput = $('textarea.js-taginput');
+
+    if (!tagInput) {
+      return true;
+    }
+
+    const tagsArr = tagInput.value.split(',').map(t => t.trim());
+
+    const errors = [];
+
+    let hasRating = false;
+    let hasSafe = false;
+    let hasOtherRating = false;
+
+    tagsArr.forEach(tag => {
+      if (ratingsTags.includes(tag)) {
+        hasRating = true;
+        if (tag === 'safe') {
+          hasSafe = true;
+        } else {
+          hasOtherRating = true;
+        }
+      }
+    });
+
+    if (!hasRating) {
+      errors.push('Tag input must contain at least one rating tag');
+    } else if (hasSafe && hasOtherRating) {
+      errors.push('Tag input may not contain any other rating if safe');
+    }
+
+    if (tagsArr.length < 3) {
+      errors.push('Tag input must contain at least 3 tags');
+    }
+
+    errors.forEach(msg => createTagError(msg));
+
+    return errors.length === 0; // true: valid if no errors
+  }
+
+  function enableUploadButton() {
+    const submitButton = $('.input--separate-top');
+    if (submitButton !== null) {
+      submitButton.disabled = false;
+      submitButton.innerText = 'Upload';
+    }
+  }
```

Comment:
This does not seem to be necessary

Reply:
It was needed because new.html.slime sets up a disable_with. I took that out of new.html.slime, but I'm not sure removing the disable_with was the best approach. I think it still needs to show "Please wait..." if the form submits fully.

Reply:
This is specifically referring to the "enable" part because the upload form isn't resettable, so it never needs to reenable the button.